tests/unit/parallel/algorithms/is_heap_tests.hpp

```diff
@@ -0,0 +1,425 @@
+
+    bool operator<(user_defined_type const& t) const
+    {
+        if (this->name < t.name)
+            return true;
+        else if (this->name < t.name)
+            return false;
+        else
+            return this->val < t.val;
+    }
+
+    const user_defined_type& operator++()
+    {
+        static const std::vector<std::string> name_list = {
+            "ABB", "ABC", "ACB", "BCA", "CAA", "CAAA", "CAAB"
+        };
+        name = name_list[std::rand() % name_list.size()];
+        ++val;
+        return *this;
+    }
+
+    std::string name;
+    int val;
+};
+
+///////////////////////////////////////////////////////////////////////////////
+template <typename ExPolicy, typename IteratorTag, typename DataType = int>
+void test_is_heap(ExPolicy policy, IteratorTag, DataType = DataType(),
+    bool test_for_is_heap = true)
+{
+    static_assert(
+        hpx::parallel::execution::is_execution_policy<ExPolicy>::value,
+        "hpx::parallel::execution::is_execution_policy<ExPolicy>::value");
+
+    typedef typename std::vector<DataType>::iterator base_iterator;
+    typedef test::test_iterator<base_iterator, IteratorTag> iterator;
+
+    std::vector<DataType> c(10007);
+    std::iota(boost::begin(c), boost::end(c), DataType(std::rand()));
+
+    auto heap_end_iter = std::next(boost::begin(c), std::rand() % c.size());
+    std::make_heap(boost::begin(c), heap_end_iter);
+
+    if (test_for_is_heap)
+    {
+        auto result = hpx::parallel::is_heap(policy,
+            iterator(boost::begin(c)), iterator(boost::end(c)));
+        auto solution = std::is_heap(std::begin(c), std::end(c));
+
+        HPX_TEST(result == solution);
+    }
+    else
+    {
+        auto result = hpx::parallel::is_heap_until(policy,
+            iterator(boost::begin(c)), iterator(boost::end(c)));
+        auto solution = std::is_heap_until(std::begin(c), std::end(c));
+
+        HPX_TEST(result.base() == solution);
+    }
+}
+
+template <typename ExPolicy, typename IteratorTag, typename DataType, typename Pred>
+void test_is_heap_with_pred(ExPolicy policy, IteratorTag, DataType, Pred pred,
+    bool test_for_is_heap = true)
+{
+    static_assert(
+        hpx::parallel::execution::is_execution_policy<ExPolicy>::value,
+        "hpx::parallel::execution::is_execution_policy<ExPolicy>::value");
+
+    typedef typename std::vector<DataType>::iterator base_iterator;
+    typedef test::test_iterator<base_iterator, IteratorTag> iterator;
+
+    std::vector<DataType> c(10007);
+    std::iota(boost::begin(c), boost::end(c), DataType(std::rand()));
+
+    auto heap_end_iter = std::next(boost::begin(c), std::rand() % c.size());
+    std::make_heap(boost::begin(c), heap_end_iter);
+
+    if (test_for_is_heap)
+    {
+        auto result = hpx::parallel::is_heap(policy,
+            iterator(boost::begin(c)), iterator(boost::end(c)), pred);
+        auto solution = std::is_heap(std::begin(c), std::end(c), pred);
+
```

Comment:
I'd suggest to use `bool` here instead of `auto`. Same amount of typing, better readability...